### PR TITLE
Explore: Pass the dataframes along with the time range of the data

### LINKED ISSUE #89039: Explore: TimeSeriesPanel crashes after switching the time range
### What happened?

TimeSeriesPanel crashes with the error below after switching from an absolute range (5 second wide) to relative "last 3 hours":

```
RangeError: Invalid array length
    at Array.push (<anonymous>)
    at i (nullInsertThreshold.ts:156:20)
    at r (nullInsertThreshold.ts:72:31)
    at h (utils.ts:132:9)
    at TimeSeriesPanel.tsx:47:32
```

The issue doesn't seem to be happening in Dashboards.

Possibly related issues:
- https://github.com/grafana/grafana/pull/79725

For more details please check: https://raintank-corp.slack.com/archives/C02TXB92G6T/p1717626732900599 

### What did you expect to happen?

The panel should display the data.

### Did this work before?

Not sure.

### How do we reproduce it?

See above

### Is the bug inside a dashboard panel?

No

### Environment (with versions)?

Grafana: Grafana v11.1.0-71799 (3a8b59f5e8)
OS: MacOS
Browser: Chrome Version 125.0.6422.142 (Official Build) (arm64)

### Grafana platform?

I use Grafana Cloud

### Datasource(s)?

Loki

### PR DESCRIPTION
**Pull request by ifrost**
Fixes #89039

```
Steps to reproduce:

1. Run a test-data metrics query with a very narrow absolute time range (spanning 1s)
2. Change the range to wide, relative range, like last 24 hours.

The update of the chart is very slow or it results an error.
```

The main reason for the bug above seems to be that for rendering a timeseries chart we use the time picker's date, not the time range that was used to request the data. Both should be basically the same but when used this way there's a timing difference of when time range and data is passed to the chart leading to:

1. Time series showing a few points for 1 second wide time frame
2. Time range changes to x-hours, the new range is passed to the chart as soon as the time range in the picker changes, but the data is stale (new data is still loading)
3. The timeline chart now needs to handle a change from a very narrow to a very wide (with lots of null data points) and seems to crash.

In theory the chart should be handle to such scenario, but anyway we should pass the data and time range together to avoid performance issues.

The fix is to actually simplify the code. We are already getting the range that was used to request the data. If we use it we can also stop having conversions between absolute and non-absolute time range.

### CODE REVIEW

File: public/app/features/alerting/unified/components/rule-editor/VizWrapper.tsx [hunk 1]
```diff
@@ -41,7 +37,7 @@ export const VizWrapper = ({ data, thresholds, thresholdsType }: Props) => {
                 eventBus={appEvents}
                 height={300}
                 width={width}
-                absoluteRange={timeRange}
+                timeRange={data.timeRange}
```
Comment (ifrost) [inline public/app/features/alerting/unified/components/rule-editor/VizWrapper.tsx:40]: @grafana/alerting-squad Please check if that won't break anything :) It looks like you're importing an Explore's component. Do you think it'd make sense to move this component somewhere to `app/core`?

**Comment by aocenas** [comment]
Is this a good change from a UX perspective? I would assume the original behavior to be more desirable. Like I have a data points `-o-o-` in a chart and then change time range so it fires a new query that may take some time, I would probably like to see at least some immediate change so that graph looks like `-------------o-o------------` (showing the new time range with the original data) while I am waiting for the additional data to load. Maybe a nitpicky thing but I would say such progressive load and UI changes are more desirable.

**Comment by ifrost** [comment]
> Is this a good change from a UX perspective? I would assume the original behavior to be more desirable. Like I have a data points -o-o- in a chart and then change time range so it fires a new query that may take some time, I would probably like to see at least some immediate change so that graph looks like -------------o-o------------ (showing the new time range with the original data) while I am waiting for the additional data to load. Maybe a nitpicky thing but I would say such progressive load and UI changes are more desirable.

Good point. It's how it works in Explore, but not in Dashboards:

Indeed in Explore the graph stretches straight away when zooming out and updates later with the data.

https://github.com/grafana/grafana/assets/745532/85073ef2-abf7-4327-8670-dda4965f600f

In Dashboards, the panel is refreshed only after getting the data:

https://github.com/grafana/grafana/assets/745532/7dab6ac7-ddcc-4f9c-897d-6f64b26cc239

So with this change Explore will be more consistent with Dashboards, though may feel less responsive. Probably much more visible when zooming in than out, like below (in Explore the zoom is immediate, though may be replaced with more detailed data after loading):

https://github.com/grafana/grafana/assets/745532/0b57b888-20bf-4450-aebb-b86e1c0d2a29

There was actually a related proposal that suggested showing an overlay loading indicator while the data is loading ([doc](https://docs.google.com/document/d/1sWlkFqzWbDVwPMAgt_tSZG3U1ar8eRyEen-PBLQq3LY/edit#heading=h.e11e7u36d67p)). 

@diegoadams any thoughts on this?

@grafana/dataviz-squad - could it be a bug in the TimeSeries panel or maybe we're doing something wrong and it should handle such use case easily?

**Comment by diegoadams** [comment]
> @diegoadams any thoughts on this?

I strongly support the idea of implementing an overlay loading indicator. This would keep users informed that the data is being fetched. @ifrost, was there a reason why this approach wasn't used originally?

**Comment by ifrost** [comment]
> I strongly support the idea of implementing an overlay loading indicator. This would keep users informed that the data is being fetched. @ifrost, was there a reason why this approach wasn't used originally?

I think the work was not prioritized and is still in the backlog.

**Review by harisrozajac — APPROVED** [review]
🚀 Looks great!

**Review by gelicia — APPROVED** [review]
Tried with test data and devenv prometheus, seems to work great!

**Comment by ifrost** [comment]
@grafana/alerting-squad please check the [comment](https://github.com/grafana/grafana/pull/90259#discussion_r1671923790) above and let me know if you have any suggestions / comments.

**Pull request merged**